apps/src/util/thumbnail.js

```diff
@@ -76,6 +80,71 @@ export function captureThumbnailFromSvg(svg) {
     .then(project.saveThumbnail);
 }
 
+/**
+ * Crops the given canvas to a smaller portion from the top-left corner.
+ * The resulting canvas will have its width and height scaled by neighborhoodThumbnailScale
+ * stored in Codebridge Registry. For now, this is only used by Python Lab neighborhood projects.
+ *
+ * @param {HTMLCanvasElement} canvas - The original canvas to crop from.
+ */
+function cropNeighborhoodCanvasFromTopLeft(canvas) {
+  const scale =
+    CodebridgeRegistry.getInstance().getNeighborhoodThumbnailScale();
+  const croppedCanvas = document.createElement('canvas');
+  croppedCanvas.width = canvas.width * scale;
+  croppedCanvas.height = canvas.height * scale;
+
+  const context = croppedCanvas.getContext('2d');
+  if (context) {
+    context.drawImage(
+      canvas,
+      0,
+      0, // Source x,y
+      canvas.width * scale,
+      canvas.height * scale, // Source width, height
+      0,
+      0, // Destination x,y
+      croppedCanvas.width,
+      croppedCanvas.height // Destination width, height
+    );
+
+    return croppedCanvas;
+  }
+}
+
+/**
+ * Converts the contents of an SVG element into an image, shrinks it to a
+ * width equal to THUMBNAIL_WIDTH preserving aspect ratio, and saves it to
+ * the server.
+ * This version is to generate thumbnail image for Python Lab projects using Neighborhood mini-app.
+ * @param {SVGElement | undefined} svg SVG element to capture the contents of.
+ */
+export function captureThumbnailFromSvgPythonlabNeighborhood(svg) {
+  if (!svg) {
+    console.warn(`Thumbnail capture failed: svg element not found.`);
+    return;
+  }
+
+  // Skip capturing a screenshot if we just captured one recently.
+  const intervalMs = Date.now() - lastCaptureTimeMs;
```

Comment:
I'm guessing this logic is carried over from previous thumbnail logic? This means that sometimes the thumbnail will be out of date for a while. If the student made a change, it saved, they make another change within a minute, it doesn't save, then they stop working, we will only have saved the first version.
Since this is longstanding behavior I think it's fine to keep, but something to consider for the future.

Reply:
Yes, I'm replicating existing logic for thumbnail generation for legacy labs. I observed before working on this update that the thumbnail is out of date for a bit. But I can understand why we want to limit the number of images stored in S3.